README
Paintown (http://paintown.sf.net)
Version 3.6.0
Jon Rafkind

Readme last updated on 10/16/2011

Topics:
  Get Source
  Compilation
    Linux
    Windows
    Cygwin
    OSX
  Install
    Linux
  Scripting
  Requirements to build
  Thanks

* Get Source:

  If you are reading this you might already have the source. If not get it from here
  Game engine: $ svn co https://paintown.svn.sourceforge.net/svnroot/paintown/trunk

  You need the svn program, which is part of subversion - http://subversion.tigris.org/, to do this. Put the data directory inside the trunk directory, trunk/data.

  As of r1124 the data/ directory is an external resource for trunk, so you only have to check out trunk and svn will automatically get data/ for you.
  
  Data: $ svn co https://paintown.svn.sourceforge.net/svnroot/paintown/data

* Compilation:
  
  --------------------------------------
  Ubuntu users: There is a script that will install the necessary packages and then
  invoke the compilation scripts.

  Type this:
  $ ./easy-compile-ubuntu
  --------------------------------------
  
  You can use scons or cmake.

  If the scons build fails because it cannot find hawknl/nl.h it may be because of this problem, http://scons.tigris.org/issues/show_bug.cgi?id=2121. This error will manifest itself like so:
  
  build/network/network.h: 5:23: error: hawknl/nl.h: No such file or directory 
  
  Try the patch listed on this page, http://scons.tigris.org/nonav/issues/showattachment.cgi/448/bug2121_proposed_patch.  
  Upgrading to scons 1.1.0 also fixes this issue. http://www.scons.org

  To change the data directory using scons, give the data= argument.
    $ scons data=my/own/path

  * Linux

    $ make
     or
    $ scons

    Will build SDL by default. Set an environment variable to get different backends.
    $ export ALLEGRO=1 # to use Allegro4 (software renderer)
    $ export ALLEGRO5=1 # to use Allegro5 (hardware renderer)

    Cmake also works.

    $ mkdir b
    $ cd b
    $ cmake ..
    $ make

  * Windows
	
    To use mingw type

      $ scons env=mingw

    Otherwise

      $ scons

    Or you can use cmake

      paintown $ md b
      paintown $ cd b
      paintown/b $ cmakesetup ..
      paintown/b $ make

    The executable will end up in bin/paintown.exe

    To build a statically linked binary give the 'static' target
    $ scons static

  * Cygwin
    
    Don't use cygwin's tools. Install python/scons/mingw for regular windows and use that.
    $ export CYGWIN=1
    $ c:\progra~1\Python24\Scripts\scons

  * OSX
    $ scons

    How to build a .dmg file of paintown:
      $ scons static
      $ misc/make-osx-bundle

    This will create paintown-x-y.app/ and paintown-x-y.dmg.

   You need buildDmg.pl: http://www.objectpark.org/buildDMG.html
   I have placed the perl script in the misc directory, misc/buildDMG.pl, so just put
   that in a directory that is part of your path.

* Install:

  * Linux

  $ ./install.sh -d <data> -b <bin>

  Copy the data directory to <data> and put a script in <bin>.
  Default <data> directory is '/usr/share/games/paintown-x.x' where x.x is the version.
  Default <bin> directory is '/usr/games'
  
  The script will prompt you to input these directories if you do not
  give -d and/or -b.

  After running install.sh (assuming default values for <data> and <bin>) you can run
  $ /usr/games/paintown

  * Anything other than Linux
    Copy the binary and data somewhere. Otherwise use the prepackaged installers.

* Scripting

  A python engine has been added to paintown so you can script a level. See
  scripting.txt for more details.

* Requirements to build:

  You need the following things to compile Paintown:
    * gcc - http://www.gnu.org/software/gcc/
    Either * scons - http://www.scons.org
           * python - http://www.python.org
    Or
           * cmake - http://www.cmake.org
    * zlib - http://www.zlib.net/
    * libpng - http://www.libpng.org/pub/png/libpng.html
    * pthreads - If you are on Unix you should have this or you can install glibc and you will get it. Windows people: http://sources.redhat.com/pthreads-win32/
    * freetype - Installed on most Unix systems along with X11. Windows people: http://gnuwin32.sourceforge.net/packages/freetype.htm.
    * Python headers, usually this comes with python
    
    One of:
      * Allegro 4.x - 4.1 or higher. http://alleg.sf.net
      * Allegro 5.x - http://alleg.sf.net
      * SDL 1.2.x - http://libsdl.org

  * Quick instructions for ubuntu (also see easy-compile-ubuntu)
    $ sudo apt-get install python-dev libfreetype6-dev liballegro4.2 zlib1g-dev

  * Detailed windows instructions
    *Note* it may be easier to get the gcc4.5 environment from svn:
      https://paintown.svn.sourceforge.net/svnroot/paintown/build-environments/gcc4.5
    Get unxutils for things like mv, rm, cp - http://www.sf.net/projects/unxutils
      Find the .zip. It unzips as usr\local\wbin, so put that in your path or just
      copy all the files in usr\local\wbin to your c:\mingw\bin directory.
    For libpng/zlib see here: http://tjaden.strangesoft.net/loadpng/mingw.html
    Python 2.4.4 (yes this is old but the game is hardcoded to use it):
      http://www.python.org/download/releases/2.4.4/
      Add c:\python24 to your PATH
      Copy libs\libpython24.a to c:\mingw\lib
    Scons - http://www.scons.org
      Then add c:\python24\scripts to your PATH
    pthreads - http://sources.redhat.com/pthreads-win32/
      get pthreads-win32-2-8-0-release.exe, let it unzip somewhere, then copy
      the stuff in prebuilt\include to your mingw\include directory. copy the
      .a files to c:\mingw\lib
    allegro - http://alleg.sf.net
      get the 4.2.3 package and dx70_mgw.zip
    Get freetype from http://gnuwin32.sourceforge.net/packages/freetype.htm. Copy
     the resulting headers into your include path. Dont copy in freetype2, just the
     path underneath it, freetype2\freetype.
    Gnu regex: http://gnuwin32.sourceforge.net/packages/regex.htm
    Use the dependancy walker: http://www.dependencywalker.com/ to check that
      all required dlls are around.

* Thanks:
  Rain sound - by inchadney. http://freesound.iua.upf.edu/usersViewSingle.php?id=28867	
  DUMB - http://dumb.sf.net
  loadpng - http://tjaden.strangesoft.net/loadpng/


doc/changelog
3.5.0 10/14/2010
-----
Paintown Engine:
 - A new move list dialog shows all the attack moves for a player
 - Local multiplayer co-op
 - Much improved joystick support

Paintown Editor:
 - Re-organized the layout
 - Onion skinning
 - The background color of the animation window can be changed
 - Additional convienence tools were added

Mugen Engine:
 - Characters and stages are discovered in the background without having
 to modify select.def
 - Many parser bugs were fixed

Menu:
 - Menu items can be customized

Other:
 - Ports to Android, PS3, Wii, Native Client
 - MP3 and Ogg support

3.4.0 12/26/2010
-----
Paintown Engine:
 - Pressing ESC brings up a menu in the game
 - Input uses events in the menu
 - New thunder and rain effects
 - Added character MJ
 - Ported backend to SDL
 - Added ability to play .nsf, .gym, etc files

Mugen Engine:
 - Parsers much improved
 - Helpers implemented
 - A few more characters work besides KFM
 - Mugen script is compiled much more efficiently
 - Tons of other bug fixes
 - Roughly 80% done

Other:
 - Mostly functional to the Wii
 - Partial port to the psp

3.3.1 2/25/2010
-----

Paintown Engine:
 - removed versus mode completely (it was still referenced by the menu)

3.3 2/21/2010
---

Paintown Engine:
 - added dynamic lights
 - network mode checks for versions
 - chatting in network mode

Paintown Menus:
 - Added Animations
 - Added Tabbed Menus
 - Clean ups removing most things from global space ie:
 - Removed option info box from global space and made it part of menu.
 - Made menu info and option info default to specific locations
 - Added static doDialog to Box
 - Use return exception in menus

Paintown Utilities:

Mugen Engine:
 - Implemented base file readers for Mugen
 - Created necessary classes to handle the Engine:
 - Fonts
 - Sprites
 - Animations
 - Sounds
 - Backgrounds
 - Menus
 - Utils
 - Characters
 - Game
 - Parser

3.2 1/11/2009
---
add stuff

3.1 8/3/2008
---
add stuff

3.0 1/28/2008
---
add stuff


src/fonts.h
/* Allegro datafile object indexes, produced by grabber v4.1.13 (WIP), Unix */
/* Datafile: /home/kazzmir/archives/scratch/c++/jezzball/PCX_fonts/fonts.dat */
/* Date: Wed Mar 31 14:12:06 2004 */
/* Do not hand edit! */

#define ANDALEMONO_PCX                   0        /* FONT */
#define BROADWAY_PCX                     1        /* FONT */
#define CALIGRAPH_PCX                    2        /* FONT */
#define CATANEO_PCX                      3        /* FONT */
#define CENTURYGOTHIC_PCX                4        /* FONT */
#define CORSIVA_PCX                      5        /* FONT */
#define FIXEDSYS_PCX                     6        /* FONT */
#define HIGHLIGHT_PCX                    7        /* FONT */
#define IRONMAIDEN_PCX                   8        /* FONT */
#define JOHNHANDY_PCX                    9        /* FONT */
#define JOKERMAN_PCX                     10       /* FONT */
#define KUNSTLER_PCX                     11       /* FONT */
#define LABAMBA_PCX                      12       /* FONT */
#define LCD_PCX                          13       /* FONT */
#define MEKANIK_PCX                      14       /* FONT */
#define METALLORD_PCX                    15       /* FONT */
#define MISTER_PCX                       16       /* FONT */
#define ODESSA_PCX                       17       /* FONT */
#define OLDDREAD_PCX                     18       /* FONT */
#define ONESTROKE_PCX                    19       /* FONT */
#define ORANGE_PCX                       20       /* FONT */
#define PARKAVENUE_PCX                   21       /* FONT */
#define PUMP_PCX                         22       /* FONT */
#define QUIXLEY_PCX                      23       /* FONT */
#define RAGE_PCX                         24       /* FONT */
#define RAUCH_PCX                        25       /* FONT */
#define SCRUFF_PCX                       26       /* FONT */
#define SMUDGER_PCX                      27       /* FONT */
#define SQUARE_PCX                       28       /* FONT */
#define STACCATO_PCX                     29       /* FONT */
#define TIRANTI_PCX                      30       /* FONT */
#define VICTORAIN_PCX                    31       /* FONT */
#define WOFACHABCER_PCX                  32       /* FONT */
#define WOFALDE_PCX                      33       /* FONT */
#define WOFAMBROS_PCX                    34       /* FONT */
#define WOFAMERUNI_PCX                   35       /* FONT */
#define WOFANDRO_PCX                     36       /* FONT */
#define WOFARGENT_PCX                    37       /* FONT */
#define WOFARIZONA_PCX                   38       /* FONT */
#define WOFARROW_PCX                     39       /* FONT */
#define WOFARTE_PCX                      40       /* FONT */
#define WOFBAIDER_PCX                    41       /* FONT */
#define WOFBARIUM_PCX                    42       /* FONT */
#define WOFBASTIA_PCX                    43       /* FONT */
#define WOFBATS_PCX                      44       /* FONT */
#define WOFBILIA_PCX                     45       /* FONT */
#define WOFBLADE_PCX                     46       /* FONT */
#define WOFBLIXA_PCX                     47       /* FONT */
#define WOFBONIX_PCX                     48       /* FONT */
#define WOFBORDER_PCX                    49       /* FONT */
#define WOFBOROLD_PCX                    50       /* FONT */
#define WOFBORSCH_PCX                    51       /* FONT */
#define WOFBURO_PCX                      52       /* FONT */
#define WOFCANDI_PCX                     53       /* FONT */
#define WOFCHIRO_PCX                     54       /* FONT */
#define WOFCLIPS_PCX                     55       /* FONT */
#define WOFCLOCK_PCX                     56       /* FONT */
#define WOFCLOCKER_PCX                   57       /* FONT */
#define WOFCOOL_PCX                      58       /* FONT */
#define WOFDELLI_PCX                     59       /* FONT */
#define WOFDRAKER_PCX                    60       /* FONT */
#define WOFEDUK_PCX                      61       /* FONT */
#define WOFELECTRO_PCX                   62       /* FONT */
#define WOFFIRNIS_PCX                    63       /* FONT */
#define WOFFITIO_PCX                     64       /* FONT */
#define WOFFLOW_PCX                      65       /* FONT */
#define WOFFUTUR_PCX                     66       /* FONT */
#define WOFGINEWRA_PCX                   67       /* FONT */
#define WOFHANDS_PCX                     68       /* FONT */
#define WOFHARDSH_PCX                    69       /* FONT */
#define WOFINTONA_PCX                    70       /* FONT */
#define WOFJOKER_PCX                     71       /* FONT */
#define WOFNEWART_PCX                    72       /* FONT */
#define ZLAST                            73       /* BMP  */

